File: trunk/compact-net/CompactNetMercFramework/MessageQueue.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace it.mintlab.mobilenet.mercframework
{
    internal class MessageQueue
    {
        private object lockToken;
        private List<Message> generalQueue;
        private Dictionary<MessageContent, List<Message>> filteredQueues;

        public MessageQueue()
        {
            lockToken = new object(); //Needed for locking
            generalQueue = new List<Message>();
            filteredQueues = new Dictionary<MessageContent, List<Message>>();
        }

        /// <summary>
        /// Enqueue a message
        /// </summary>
        /// <param id="message"></param>
        public void enqueue(Message message)
        {
            lock (lockToken)
            {
                generalQueue.Add(message);
                MessageContent key = message.getContent();
                if (filteredQueues.ContainsKey(key))
                {
                    filteredQueues[key].Add(message);
                }
                else
                {
                    List<Message> l = new List<Message>();
                    l.Add(message);
                    filteredQueues.Add(key, l);
                }
            }
        }

        /// <summary>
        /// Dequeue a message
        /// </summary>
        /// <returns></returns>
        public Message dequeue()
        {
            lock (lockToken)
            {
                if (generalQueue.Count == 0)
                    throw new InvalidOperationException("MessageQueue empty");
                Message result = generalQueue[0];
                generalQueue.RemoveAt(0);
                MessageContent key = result.getContent();
                filteredQueues[key].RemoveAt(0);
                if (filteredQueues.Count == 0)
                    filteredQueues.Remove(key);
                return result;
            }
        }

        /// <summary>
        /// Dequeue a message of a particular filter
        /// </summary>
        /// <returns></returns>
        public Message dequeue(MessageContent filter)
        {
            lock (lockToken)
            {
                if (!filteredQueues.ContainsKey(filter))
                    throw new InvalidOperationException("MessageQueue empty for content "+filter);
                List<Message> l = filteredQueues[filter];
                Message result = l[0];
                l.RemoveAt(0);
                if (filteredQueues.Count == 0)
                    filteredQueues.Remove(filter);
                generalQueue.Remove(result);
                return result;
            }
        }

        /// <summary>
        /// Returns true if this queue contains the specified filter
        /// </summary>
        /// <param id="filter">filter whose presence in this queue is to be tested</param>
        /// <returns>true if the specified element is present; false otherwise</returns>
        public bool contains(MessageContent filter)
        {
            return filteredQueues.ContainsKey(filter);
        }


        /// <summary>
        /// Clear all messages
        /// </summary>
        public void clear()
        {
            lock (lockToken)
            {
                generalQueue.Clear();
                filteredQueues.Clear();
            }
        }

        /// <summary>
        /// Clear all messages of a filter declared
        /// </summary>
        /// <param id="filter"></param>
        public void clear(MessageContent filter)
        {
            lock (lockToken)
            {
                if (!filteredQueues.ContainsKey(filter)) return;
                List<Message> l = filteredQueues[filter];
                filteredQueues.Remove(filter);
                foreach (Message msg in l)
                {
                    generalQueue.Remove(msg);
                }
            }
        }

        /// <summary>
        /// Return True if queue is empty
        /// </summary>
        /// <returns></returns>
        public bool isEmpty()
        {
            if (generalQueue.Count == 0)
                return true;
            else
                return false;
        }

    }
}

File: java/JavaMercFramework/src/main/it/mintlab/desktopjava/mercframework/MessageQueue.java
package it.mintlab.desktopjava.mercframework;

	public class MessageQueue
	{
		private Object lockToken;
		private java.util.ArrayList<Message> generalQueue;
		private java.util.HashMap<MessageContent, java.util.ArrayList<Message>> filteredQueues;

		public MessageQueue()
		{
			lockToken = new Object(); //Needed for locking
			generalQueue = new java.util.ArrayList<Message>();
			filteredQueues = new java.util.HashMap<MessageContent, java.util.ArrayList<Message>>();
		}

		/** 
		 Enqueue a message
		 
		 <param id="message">
		*/
		public final void enqueue(Message message)
		{
			synchronized (lockToken)
			{
				generalQueue.add(message);
				MessageContent key = message.getContent();
				if (filteredQueues.containsKey(key))
				{
					filteredQueues.get(key).add(message);
				}
				else
				{
					java.util.ArrayList<Message> l = new java.util.ArrayList<Message>();
					l.add(message);
					filteredQueues.put(key, l);
				}
			}
		}

		/** 
		 Dequeue a message
		 
		 @return 
		 * @throws Exception 
		*/
		public final Message dequeue() throws Exception
		{
			synchronized (lockToken)
			{
				if (generalQueue.isEmpty())
				{
					throw new Exception("MessageQueue empty");
				}
				Message result = generalQueue.get(0);
				generalQueue.remove(0);
				MessageContent key = result.getContent();
				filteredQueues.get(key).remove(0);
				if(filteredQueues.size() == 0)
					filteredQueues.remove(key);
				return result;
			}
		}

		/** 
		 Dequeue a message of a particular filter
		 
		 @return 
		 * @throws Exception 
		*/
		public final Message dequeue(MessageContent filter) throws Exception
		{
			synchronized (lockToken)
			{
				if (!filteredQueues.containsKey(filter))
				{
					throw new Exception("MessageQueue empty for content "+filter);
				}
				java.util.ArrayList<Message> l = filteredQueues.get(filter);
				Message result = l.get(0);
				l.remove(0);
				if(filteredQueues.size() == 0)
					filteredQueues.remove(filter);
				generalQueue.remove(result);
				return result;
			}
		}

		/** 
		 Returns true if this queue contains the specified filter
		 
		 <param id="filter">filter whose presence in this queue is to be tested
		 @return true if the specified element is present; false otherwise
		*/
		public final boolean contains(MessageContent filter)
		{
			return filteredQueues.containsKey(filter);
		}


		/** 
		 Clear all messages
		 
		*/
		public final void clear()
		{
			synchronized (lockToken)
			{
				generalQueue.clear();
				filteredQueues.clear();
			}
		}

		/** 
		 Clear all messages of a filter
		 
		 <param id="category">
		*/
		public final void clear(MessageContent filter)
		{
			synchronized (lockToken)
			{
				if (!filteredQueues.containsKey(filter))
				{
					return;
				}
				java.util.ArrayList<Message> l = filteredQueues.get(filter);
				filteredQueues.remove(filter);
				for (Message msg : l)
				{
					generalQueue.remove(msg);
				}
			}
		}

		/** 
		 Return True if queue is empty
		 
		 @return 
		*/
		public final boolean isEmpty()
		{
			if (generalQueue.isEmpty())
			{
				return true;
			}
			else
			{
				return false;
			}
		}

	}